### DISCUSSION TITLE
Keycloak freezes after cache warnings

### DISCUSSION THREAD

**Discussion by knlambert** [keycloak/keycloak#16405] [Q&A]
Hello,

We are running Keycloak `16.1.0` on jboss wildfly, in HA.
We also have some extensions with custom business logic that we added (mostly custom required user actions, and some REST apis).

Since the end of december, the instances began to freeze after maybe 2-3 days of uptime: the requests are not served anymore for an amount of time between 2s and 10s, if not more. Eventually the pods end up being killed by Kubernetes, causing a service interruption.

The issue is always there, it's just that it usually starts quiet, with a small amount of freeze / warning logs once in a while, until it eventually becomes too slow for the liveness probe, and that's the moment Kubernetes kills the pods.

This is the kind of warnings in the logs everytime it happens:

```bash
INFO 2023-01-11T14:57:02.553037841Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,545 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4419) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.542485905Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,540 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4425) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.542434074Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,537 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4417) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.435943052Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,435 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4461) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.435315377Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,435 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4458) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.432246646Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,431 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4457) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.397334729Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,397 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4453) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.396942757Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,396 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4411) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.384986127Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,384 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4452) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.382829017Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,382 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4447) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.280876883Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,280 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4438) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.273621340Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,273 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4412) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.160630064Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,160 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4398) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.154460806Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,154 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4443) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
INFO 2023-01-11T14:57:02.077621845Z [resource.labels.containerName: keycloak] [0m[33m14:57:02,077 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4431) Failed to replace entity 'cfd71a5c-a734-4aa8-b394-d5220a3bc279' in cache 'sessions'
[...]
```
Sometimes, errors with the 'clientSessions' cache are in the mix too:

```bash
[0m[33m09:55:22,691 WARN [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (default task-4069) Failed to replace entity 'be837174-9957-435b-927f-1c1836f62601' in cache 'clientSessions'
```

I don't know if this is relevant for the issue, but when it arrives, I trace the session using the REST API, and I always end up with 3 specific users, even after a restart. These users have been there for a long time though, and don't seem different at first glance.

So my questions would be :
* Do we have a way to have a better error message, and what is keycloak trying to do when this happens ? I think a better understanding of what is happening under the hood may help.
* Any idea of what the source of the error could be, or of how we could trace it ? (looking for a specific kind of debug log maybe ?)

Thank you.

**Comment by LeshkaSafronov** [comment]
Hello,

I also see the described problem (Keycloak 20.0.1). 

Step to reproduce: make load testing of the same refresh token.
Env: docker instance in docker compose, regular realm without any customisations, load: ~50 threads.

After a while latency starting to increase and warn messages:
"2023-01-18 20:28:54,468 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-90) Failed to replace entity '90150832-8805-4722-9228-b4be143667df' in cache 'sessions'" 
 appears.

**Reply by sschu** [reply]
Do you have any metrics of cache sizes and memory consumption?

**Reply by joseldmc** [reply]
Hi  @sschu, I work with @knlambert and these are some of the metrics we have for the keycloak memory:

As you can see in the chart below, it never seems to go higher than 2Gb
![image](https://user-images.githubusercontent.com/9178317/216215330-843481a9-9af0-4244-989f-51cfe3766737.png)

That is expected because the maximum memory the JVM heap will use is 1gb per pod. The pod itself is configured with a limit of 2Gi to allow room for some non-heap memory.
![image](https://user-images.githubusercontent.com/9178317/216214725-9871d079-5eeb-4959-a406-dfae5dfe3a0e.png)

**Comment by kauehmoreno** [comment]
Is there any solution to that error? Im facing the same issue using keykloack 21.1.1... And My test is login and refreshing token to see that value registered inside the cache(Im using an external one to control it), and I have been seen that warning log everytime.

**Comment by sirishkumar** [comment]
We had similar issue for the first time. We are using Keycloak 22.0.5 with following configuration for infinispan with 5 pods in the Keycloak Kubernetes statefulset.

```
keycloak-1 keycloak 2024-02-07 15:43:20,045 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-5185) Failed to replace entity 'e66492b4-9f86-4b53-b675-2b41b281aa03' in cache 'sessions'
keycloak-1 keycloak 2024-02-07 15:43:20,090 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-5184) Failed to replace entity 'e66492b4-9f86-4b53-b675-2b41b281aa03' in cache 'sessions'
keycloak-1 keycloak 2024-02-07 15:43:20,283 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-5188) Failed to replace entity 'e66492b4-9f86-4b53-b675-2b41b281aa03' in cache 'sessions'
keycloak-1 keycloak 2024-02-07 15:43:20,307 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-5190) Failed to replace entity 'e66492b4-9f86-4b53-b675-2b41b281aa03' in cache 'sessions'
keycloak-1 keycloak 2024-02-07 15:43:20,387 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-5189) Failed to replace entity 'e66492b4-9f86-4b53-b675-2b41b281aa03' in cache 'sessions'
keycloak-1 keycloak 2024-02-07 15:43:20,484 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-5193) Failed to replace entity 'e66492b4-9f86-4b53-b675-2b41b281aa03' in cache 'sessions'
keycloak-1 keycloak 2024-02-07 15:43:20,583 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-5177) Failed to replace entity '3eff39d6-fe4e-4f70-ba9a-445be842270a' in cache 'clientSessions'
```

```
<distributed-cache name="sessions" owners="2" statistics="true">
    <expiration lifespan="-1"/>
</distributed-cache>
```

Session lifespan settings in Keycloak
<img width="651" alt="image" src="https://github.com/keycloak/keycloak/assets/27522/9aa4e37b-705f-4726-b9e5-ea368c06f08c">

Session statistics below during the outage:

<img width="3348" alt="image" src="https://github.com/keycloak/keycloak/assets/27522/596208ec-80f7-430d-97df-5fabace898cf">

Memory:
Both Heap and Non-Heap memory consumption was way below configured limits. All the pods where consuming around 1 GB of memory of configured 8 GB limit.

**Reply by wolfganghorner** [reply]
I´m experiencing the same problem with Keycloak version 22.0.5 when running 2 keycloak pods.  After about 1 day the problem starts with such warnings in the server log and later no OIDC login is working anymore. Scaling down to 1 pod solves the problem.
Did anyone already found a solution for this?
2024-04-16 12:03:03,786 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-1728) Failed to replace entity '231a5e95-ec0b-4f97-a858-405964a4e420' in cache 'sessions'
2024-04-16 12:03:04,964 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-1716) Failed to replace entity '231a5e95-ec0b-4f97-a858-405964a4e420' in cache 'sessions'
2024-04-16 12:03:05,505 WARN  [org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction] (executor-thread-1719) Failed to replace entity '231a5e95-ec0b-4f97-a858-405964a4e420' in cache 'sessions'

**Reply by shawkins** [reply]
@sirishkumar @wolfganghorner this discussion spans quite a bit of time / versions, so the recommendation would be to replicate this behavior on as new of a keycloak version as possible, and provide logs along with a thread dump (if the instance seems fully frozen) as part of a new issue.
